### DISCUSSION TITLE
Question: How to set YUV420p byte array as a source for MediaPlayerElement in WinUI3

### DISCUSSION THREAD

**Discussion by MayankKr10** [microsoft/microsoft-ui-xaml#8654] [Q&A]
I have byte arrays for a video stream in YUV420p format. I want to set this as the source for MediaPlayerElement to play the video in WinUI3 application. What is the best (and fastest) way to do this?
Note: I do NOT want to save it in a file since it is a real-time video to be rendered inside the app and we want to avoid any unnecessary memory/time consumption

I came across MediaElement as well, but the only solutions I found included reading the data from a file (which we want to avoid since it is a real-time video feed) unlike Silverlight. Also, I am not sure if MediaElement is actually supported in WinUI3

**Comment by brabebhin** [comment]
You need to use a MediaStreamSource. You can check out https://github.com/ffmpeginteropx/FFmpegInteropX if you want an advanced implementation.

**Comment by MayankKr10** [comment]
I have tried the library you mentioned but it is not compatible with our app.
Alternatively, this is what I have tried right now, although it's not working for YUV420 (works fine for mp4 if I change the MIME type accordingly):
            
            InMemoryRandomAccessStream randomAccessStream = new InMemoryRandomAccessStream();
            _ = randomAccessStream.WriteAsync(arr.AsBuffer());
            randomAccessStream.Seek(0);

            MediaSource stream1 = MediaSource.CreateFromStream(randomAccessStream, "video/x-oggyuv"); // works with "video/mp4" for mp4 stream
            _ = stream1.OpenAsync();
            VideoPlayer.Source = stream1;

Here's a reference image for the error on integration the aforementioned library-
![image](https://user-images.githubusercontent.com/48569729/211322709-1d868b17-c8d6-4752-946e-afbcb0e70f6d.png)

**Comment by brabebhin** [comment]
On the other repo you mentioned WPF, here your app seems to be winUI.
For WPF, I responded on the other repo, for winUI, there's a sample app in the repo which is winUI and works properly, you can follow that and see how to integrate the framework there.

The reason why you can't pass YUV directly to the MediaSource is that windows doesn't decode YUV directly, it needs a converter. This is why you need a MediaStreamSource (not a MediaSource - I know, pretty confusing naming).
